Webhook provider (#3063)
kubernetes-sigs/external-dns: docs/tutorials/plugin-provider.md

# Plugin provider

> The Plugin provider is a provider that allows DNS providers to integrate with ExternalDNS via an HTTP interface. The Plugin provider implements the Provider interface but instead of implementing code specific to a provider, it implements and HTTP client that sends request to an HTTP API. The idea behind this is that providers can be implemented in a separate program that exposes an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars of the ExternalDNS container, but this is not strictly a requirement.

Reviewer: ```suggestion
The Plugin provider is a provider that allows DNS providers to integrate with ExternalDNS via an HTTP interface. The Plugin provider implements the Provider interface but instead of implementing code specific to a provider, it implements an HTTP client that sends request to an HTTP API. The idea behind this is that providers can be implemented in a separate program that exposes an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars of the ExternalDNS container and listen on localhost only, but this is not strictly a requirement.
Be aware that the Plugin provider client does neither support authentication nor authorization, so exposing the plain Plugin provider over the network can put the integrity of your DNS records at risk.  
```

I feel like there should be a stronger message here as this could become a security issue quickly, if the user is holding it wrong.

Reviewer 2: Yeah localhost is good because we do not need to think about authnz nor secret rotation.

Author: You are right, this could become problematic, let me add it to the text.

> The Plugin provider is a provider that allows DNS providers to integrate with ExternalDNS via an HTTP interface. The Plugin provider implements the Provider interface but instead of implementing code specific to a provider, it implements and HTTP client that sends request to an HTTP API. The idea behind this is that providers can be implemented in a separate program that exposes an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars of the ExternalDNS container and listen on localhost only, but this is not strictly a requirement even if we do not recommend other setups.

Reviewer: ```suggestion
The provider "plugin" allows to integrate ExternalDNS with DNS providers via an HTTP interface.
The provider "plugin" implements the Provider interface. Instead of implementing code specific to a provider, it implements an HTTP client that sends request to an HTTP API. 
The idea behind this is, that providers can be implemented in a separate program. These programs expose an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars in the same pod as the ExternalDNS container and listen on localhost only. This is not strictly a requirement, still we would not recommend other setups.
```

Tried to make this a bit more readable by shortening the sentences.

Author: Sounds legit, but I will refactor this instead of committing it as I want to have some changes.

## Architectural diagram

> ![Pluging provider](../img/plugin-provider.png)

Reviewer: ```suggestion
![Plugin provider](../img/plugin-provider.png)
```

Author: Dang.

## API guarantees

> Providers implementing the HTTP API have to keep in sync with changes to the Go types `plan.Changes` and `endpoint.Endpoint`. We do not expect to make significant changes to those types given the maturity of the project, but we can't exclude that changes will need to happen. We commit to publishing changes to those in the release notes, to ensure that providers implementing the API can keep providers up to date quickly.

Reviewer: Have you considered introducing a versioning of the endpoint the plugin provider (e.g. localhost:8888/v1/dns) is hitting to be able to provide a compatibility layer?

Reviewer 2: As far as I know API's have better ways of handling versioning than a path.
http://opensource.zalando.com/restful-api-guidelines/#compatibility and especially
http://opensource.zalando.com/restful-api-guidelines/#114

Author: @mrueg good point, I was never a fan of versions in the URL. What @szuecs is proposing makes sense to me (especially having worked with those API guidelines before) so I am happy to implement that.

Reviewer: Thanks for sharing @szuecs! Negotiating by content-type sounds good to me as well.

## Provider registry

> To simplify the discovery of providers, we will accept pull requests that will add links to providers in the README file. This list will serve the only purpose of simplifying finding providers and will not constitute an official endorsment of any of the externally implemented providers unless otherwise specified.

Reviewer: ```suggestion
To simplify the discovery of providers, we will accept pull requests that will add links to providers in the [README](../../README.md) file. This list will serve the only purpose of simplifying finding providers and will not constitute an official endorsment of any of the externally implemented providers unless otherwise specified.
```

The idea behind it is that providers can be implemented in a separate program: these programs expose an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars in the same pod of the ExternalDNS container and listen on localhost only. This is not strictly a requirement, but we would not recommend other setups.

## Architectural diagram

![Plugin provider](../img/plugin-provider.png)

## API guarantees

Providers implementing the HTTP API have to keep in sync with changes to the Go types `plan.Changes` and `endpoint.Endpoint`. We do not expect to make significant changes to those types given the maturity of the project, but we can't exclude that changes will need to happen. We commit to publishing changes to those in the release notes, to ensure that providers implementing the API can keep providers up to date quickly.

## Implementation requirements

The following table represents the methods to implement mapped to their HTTP method and route.

| Provider method | HTTP Method | Route |
| --- | --- | --- |
| Records | GET | /records |
| ApplyChanges | POST | /records |
| PropertyValuesEqual | POST | /propertyvaluesequal |

> | AdjustEndpoints | POST | /adjustendpoints |

Reviewer 3: I didn't find an API specification for this HTTP API – I guess what you have here is it (plus the definition of the Go interface with the data structures, assuming standard serialization)?

I didn't find an actual documentation of these methods (i.e. what they are supposed to do), but from what I could gather, it's something like this:

* Records – returns all the records which are currently defined.
* ApplyChanges – sends a list of changes to the Records to be applied.
* PropertyValuesEqual – updates a single property. I guess that's for configuration changes?
* AdjustEndpoints – ??? (takes a list of endpoints, returns a similar list)

I'd really recommend writing this down somewhere (at least as comments in provider.go).

From this, I think ApplyChanges should not become a POST, but a PATCH request to /records, as it contains a list of changes to the existing /records resource.
PropertyValuesEqual sounds like it also should be a PATCH to some configuration resource (maybe `PATCH /properties`).
I have no clue what `AdjustEndpoints` is supposed to do, so I can't really recommend something here.

Reviewer 4: To implement a provider, you need to understand the `Provider` interface. This goes regardless of whether you're implementing it embedded or implementing the HTTP protocol.

That the documentation of the `Provider` interface is incomplete is a separate issue. For now you need to read the code.

The ApplyChanges call isn't atomic, so PATCH isn't appropriate. I think the current PR is correct in representing it as a remote procedure call.

Additionally, the server needs to respond to `GET` requests on `/` to negotiate versions by content type as described in [this document](http://opensource.zalando.com/restful-api-guidelines/#114). The server needs to respond to those requests by reading the `Accept` header and responding with a corresponding `Vary` header including the value `Content-Type` and a `Content-Type` header specifying the supported media type format and version.

> **NOTE**: all responses sent with a status code different from `20X` will be considered a failure on ExternalDNS' side.

Reviewer 2: Maybe we should elaborate on 4xx and 5xx and other errors. What does the client do?
5xx -> should be retried
4xx -> log error or fatal?
3xx -> ?
2yx -> y != 0 ??

> **NOTE**: all responses sent with a status code different from `20X` will be considered a failure on the ExternalDNS' side.

Reviewer: ```suggestion
**NOTE**: all responses sent with a status code different from `20X` will be considered a failure on ExternalDNS' side.
```

Providers implementing the HTTP API have to keep in sync with changes to the Go types `plan.Changes` and `endpoint.Endpoint`. We do not expect to make significant changes to those types given the maturity of the project, but we can't exclude that changes will need to happen. We commit to publishing changes to those in the release notes, to ensure that providers implementing the API can keep providers up to date quickly.

## Implementation requirements

The following table represents the methods to implement mapped to their HTTP method and route.

| Provider method | HTTP Method | Route |
| --- | --- | --- |
| Records | GET | /records |
| ApplyChanges | POST | /records |

> | PropertyValuesEqual | GET | /propertiesvaluesequal |

Reviewer: ```suggestion
| PropertiesValuesEqual | GET | /propertiesvaluesequal |
```

Author: Rejecting as the name of the interface is `PropertyValuesEqual`.

Reviewer: Ah understood! Thanks!
